File: apps/movil/lib/dominio/configuracion.dart
/// `validarGateway` — el equivalente Dart de `validarConfiguracion`
/// (`packages/dominio-cliente/src/configuracion.ts`, D-A6). Puro, sin Flutter y sin
/// `dart:io`: no depende de la plataforma, así se prueba igual en `flutter test` que
/// dentro de la app.
///
/// Historia: `dominio/cliente.dart` tenía un valor por omisión
/// (`http://localhost/api/v1`) para `baseDelGateway`. En un build de release sin
/// `--dart-define=API=...`, ese valor por omisión es **el propio teléfono de quien
/// instaló la app** — el mismo hallazgo que los `gateway.ts` de backoffice y web.
library;

/// El resultado: o la URL usable, o el motivo del rechazo — nunca las dos cosas, y
/// nunca una excepción (la función no lanza).
class ResultadoValidacionGateway {
  const ResultadoValidacionGateway.valida(this.url)
    : valida = true,
      motivo = null;

  const ResultadoValidacionGateway.invalida(this.motivo)
    : valida = false,
      url = null;

  final bool valida;
  final String? url;
  final String? motivo;

  @override
  String toString() => valida
      ? 'ResultadoValidacionGateway.valida($url)'
      : 'ResultadoValidacionGateway.invalida($motivo)';
}

final RegExp _rutaVersionada = RegExp(r'/v\d+/?$');
const Set<String> _loopback = {'localhost', '127.0.0.1', '::1', '0.0.0.0'};

/// Valida la URL del gateway para el build de la app.
///
/// - `release`: el equivalente de "producción" — en `debug`/`profile` se admite el
///   simulado (Prism) en `localhost`/`10.0.2.2` (el alias del emulador de Android a la
///   máquina anfitriona); en `release`, nunca.
/// - `hostsPermitidos`: la lista compilada de hosts propios (D-A6) — infra la carga en
///   el despliegue (Q-J4). En `release`, sin ningún host en la lista, **todo** host
///   externo queda rechazado — no hay "origen propio" implícito como en un navegador.
ResultadoValidacionGateway validarGateway(
  String? url, {
  required bool release,
  List<String> hostsPermitidos = const [],
}) {
  final cruda = (url ?? '').trim();
  if (cruda.isEmpty) {
    return const ResultadoValidacionGateway.invalida('vacia');
  }
  if (cruda.startsWith('//')) {
    return const ResultadoValidacionGateway.invalida(
      'esquema-relativo-a-protocolo',
    );
  }

  Uri analizada;
  try {
    analizada = Uri.parse(cruda);
  } on FormatException {
    return const ResultadoValidacionGateway.invalida('formato-invalido');
  }
  if (analizada.scheme != 'http' && analizada.scheme != 'https') {
    return const ResultadoValidacionGateway.invalida('esquema-no-soportado');
  }
  if (analizada.host.isEmpty) {
    return const ResultadoValidacionGateway.invalida('formato-invalido');
  }

  final host = analizada.host.toLowerCase();
  final esLoopback = _loopback.contains(host) || host.startsWith('127.');
  if (esLoopback && release) {
    return const ResultadoValidacionGateway.invalida('localhost-no-permitido');
  }

  final permitidos = hostsPermitidos
      .map((h) => h.trim().toLowerCase())
      .where((h) => h.isNotEmpty)
      .toSet();
  final hostPermitido = permitidos.contains(host);

  if (!hostPermitido && !esLoopback) {
    return const ResultadoValidacionGateway.invalida('host-ajeno');
  }
  if (release && analizada.scheme != 'https' && !esLoopback) {
    return const ResultadoValidacionGateway.invalida('sin-tls-en-produccion');
  }

  var ruta = analizada.path;
  if (ruta.length > 1 && ruta.endsWith('/')) {
    ruta = ruta.substring(0, ruta.length - 1);
  }
  if (!_rutaVersionada.hasMatch(ruta)) {
    return const ResultadoValidacionGateway.invalida('sin-ruta-versionada');
  }

  var normalizada = cruda;
  if (normalizada.length > 1 && normalizada.endsWith('/')) {
    normalizada = normalizada.substring(0, normalizada.length - 1);
  }
  return ResultadoValidacionGateway.valida(normalizada);
}

File: packages/dominio-cliente/src/configuracion.ts
/**
 * `validarConfiguracion` — la función pura y compartida que decide si una URL de
 * gateway es utilizable, sin efectos y sin depender de ningún framework (D-A6).
 *
 * Historia: los tres `gateway.ts` (backoffice, web, y su equivalente en la app) caían
 * en silencio a `http://localhost:4010/api/v1` cuando faltaba la etiqueta/variable de
 * configuración. En el navegador de un operador, "localhost" es **su propia máquina**:
 * un despliegue mal configurado terminaba mandando tráfico financiero a donde sea que
 * corra el navegador, sin ningún aviso. Esta función es el único lugar donde se decide
 * qué es una configuración válida; los `gateway.ts` de cada app solo la llaman y, si
 * es inválida, devuelven vacío — **nunca** un valor por omisión que apunte a algún host.
 *
 * Reglas (D-A6):
 *  1. En producción, se acepta una ruta relativa (mismo origen implícito) o una URL
 *     absoluta de exactamente el mismo origen que `origenPropio` (con TLS).
 *  2. `localhost` y `127.0.0.1` (o cualquier variante de loopback) nunca se aceptan,
 *     salvo en modo `desarrollo` — ahí es Prism, el simulado local (`yarn dev:mock`).
 *  3. La ruta debe terminar en un segmento de versión (`/v1`, `/v2`, ...): sin eso no
 *     hay contrato al que apuntar.
 *  4. Fuera del origen propio, solo se acepta un host de la lista explícita
 *     `hostsPermitidos` (quién la carga: infra, en el despliegue — Q-J4).
 */

/** Los tres modos posibles en los que se resuelve el gateway. `demo` es *dato*, no red: ver `proveedor-fuentes.ts`. */
export type ModoGateway = 'produccion' | 'desarrollo'

export type OpcionesValidacionGateway = {
  /** `produccion`: TLS y mismo origen (o lista) obligatorios. `desarrollo`: se acepta loopback (Prism). */
  readonly modo: ModoGateway
  /** El origen del propio front (`location.origin` en el navegador; la URL pública del sitio en SSR). */
  readonly origenPropio: string
  /** Hosts ajenos al origen propio explícitamente permitidos (infra los carga en el despliegue). */
  readonly hostsPermitidos?: readonly string[]
}

export type ResultadoValidacionGateway =
  | { readonly valida: true; readonly url: string }
  | { readonly valida: false; readonly motivo: MotivoRechazo }

export type MotivoRechazo =
  | 'vacia'
  | 'esquema-no-soportado'
  | 'esquema-relativo-a-protocolo'
  | 'localhost-no-permitido'
  | 'sin-tls-en-produccion'
  | 'host-ajeno'
  | 'sin-ruta-versionada'
  | 'formato-invalido'

const ESQUEMAS_SOPORTADOS = new Set(['http:', 'https:'])
const LOOPBACK = new Set(['localhost', '127.0.0.1', '::1', '0.0.0.0'])
const RUTA_VERSIONADA = /\/v\d+\/?$/

function sinBarraFinal(ruta: string): string {
  return ruta.length > 1 && ruta.endsWith('/') ? ruta.slice(0, -1) : ruta
}

function normalizarOrigen(origen: string): string {
  try {
    const u = new URL(origen)
    return `${u.protocol}//${u.host}`.toLowerCase()
  } catch {
    return origen.trim().toLowerCase().replace(/\/$/, '')
  }
}

function normalizarHosts(hosts: readonly string[] | undefined): readonly string[] {
  return (hosts ?? []).map((h) => h.trim().toLowerCase()).filter((h) => h.length > 0)
}

function esLoopback(hostname: string): boolean {
  const h = hostname.toLowerCase()
  return LOOPBACK.has(h) || h.startsWith('127.')
}

export function validarConfiguracion(url: string | null | undefined, opciones: OpcionesValidacionGateway): ResultadoValidacionGateway {
  const cruda = (url ?? '').trim()
  if (cruda.length === 0) return { valida: false, motivo: 'vacia' }

  // "//otro-host/..." hereda el protocolo de la página: nunca se acepta a ciegas, sea
  // cual sea el origen — es la forma que un atacante usa para camuflar un host ajeno.
  if (cruda.startsWith('//')) return { valida: false, motivo: 'esquema-relativo-a-protocolo' }

  // Ruta relativa: mismo origen implícito. Solo se exige la versión al final.
  if (cruda.startsWith('/')) {
    const ruta = sinBarraFinal(cruda)
    if (!RUTA_VERSIONADA.test(ruta)) return { valida: false, motivo: 'sin-ruta-versionada' }
    return { valida: true, url: ruta }
  }

  let analizada: URL
  try {
    analizada = new URL(cruda)
  } catch {
    return { valida: false, motivo: 'formato-invalido' }
  }

  if (!ESQUEMAS_SOPORTADOS.has(analizada.protocol)) return { valida: false, motivo: 'esquema-no-soportado' }

  const hostname = analizada.hostname.toLowerCase()
  if (esLoopback(hostname) && opciones.modo !== 'desarrollo') return { valida: false, motivo: 'localhost-no-permitido' }

  const origenPropioNormalizado = normalizarOrigen(opciones.origenPropio)
  const origenDeUrl = `${analizada.protocol}//${analizada.host}`.toLowerCase()
  const esMismoOrigen = origenPropioNormalizado.length > 0 && origenDeUrl === origenPropioNormalizado
  const hostsPermitidos = normalizarHosts(opciones.hostsPermitidos)
  const esHostPermitido = hostsPermitidos.includes(hostname)
  const esLoopbackEnDesarrollo = esLoopback(hostname) && opciones.modo === 'desarrollo'

  if (opciones.modo === 'produccion' && analizada.protocol !== 'https:' && !esLoopbackEnDesarrollo) {
    return { valida: false, motivo: 'sin-tls-en-produccion' }
  }

  if (!esMismoOrigen && !esHostPermitido && !esLoopbackEnDesarrollo) {
    return { valida: false, motivo: 'host-ajeno' }
  }

  const ruta = sinBarraFinal(analizada.pathname)
  if (!RUTA_VERSIONADA.test(ruta)) return { valida: false, motivo: 'sin-ruta-versionada' }

  return { valida: true, url: sinBarraFinal(cruda) }
}
